### DISCUSSION TITLE
Device code grant flow

### DISCUSSION THREAD

**Discussion by 2547techno** [Chatterino/chatterino2#5169] [Ideas]
Implement device code grant flow in chatterino so refresh tokens can be used in the client, no longer needing users to re-auth every x days due to the current use of implicit grant flow

**Comment by iProdigy** [comment]
One benefit of the re-auths is we can augment the scopes on the token (as twitch releases new APIs we want to use)

If we switched to device flow, we would also have to start tracking which scopes are required for the installed c2 version (and compare with the scopes on the stored token)

**Reply by 2547techno** [reply]
good point

**Reply by 2547techno** [reply]
just a shower thought:

the issue of a user updating chatterino where new functionalities are added that need new scopes is already an issue current users may face, this won't be an issue exclusive to device flow, albeit possibly a more prolonged issue. Since users are forced to re-auth every 90 days, there is a possibility that they will not use said new functionality before the next re-auth, but also a possibility they might (maybe higher than the former since they want to try out new features).

I haven't checked this, but I assume chatterino already handles the case where the currently used token does not have enough scopes by printing an error message in chat instructing the user to re-auth, in the case of device flow this wouldn't need to change and needing to track required scopes with c2 version does not solve a new issue, rather (a version of) an already existing one.

**Comment by pajlada** [comment]
Currently a bit complex to implement since the lifetime of the access-token is short, but it's probably where we want to move at some point

**Comment by Nerixyz** [comment]
I made a super basic implementation during the beta and rebased it now: https://github.com/Nerixyz/chatterino2/tree/experiment/device-auth. [Correctly] refreshing the token is probably the hardest task here (e.g. make sure the token is refreshed before we connect to IRC/PubSub/EventSub).

https://github.com/Chatterino/chatterino2/assets/19953266/259ba236-3f02-4375-9ef9-c5cc455ebdd9
